infer/resource.go

```diff
@@ -211,54 +214,127 @@ type fieldGenerator struct {
-	alwaysSecret map[string]bool
+	alwaysSecret bool
 	// The set of tags that should never be secret
-	neverSecret map[string]bool
-	// A map from a field to its dependencies
-	deps map[string][]string
+	neverSecret bool
+	// A map from a field to its dependencies.
+	deps []string
+
+	// If the output is known, regardless of other factors.
+	known bool
 }
 
 // MarkMap mutates m to comply with the result of the fieldGenerator, applying
 // computedness and secretness as appropriate.
-func (g *fieldGenerator) MarkMap(inputs, m resource.PropertyMap) {
-	// Flow secretness and computedness
-	for output, inputList := range g.deps {
-		output := resource.PropertyKey(output)
-		for _, input := range inputList {
-			input := inputs[resource.PropertyKey(input)]
-			if input.IsComputed() && !m[output].IsComputed() {
-				m[output] = resource.MakeComputed(m[output])
-				break
-			}
-			if input.IsSecret() && !m[output].IsSecret() {
-				m[output] = resource.MakeSecret(m[output])
-			}
+func (g *fieldGenerator) MarkMap(isCreate, isPreview bool) func(oldInputs, inputs, m resource.PropertyMap) {
+	return func(oldInputs, inputs, m resource.PropertyMap) {
+		// Flow secretness and computedness
+		for k, v := range m {
+			m[k] = markField(g.getField(string(k)), k, v, oldInputs, inputs, isCreate, isPreview)
 		}
 	}
+}
+
+func markComputed(
+	field *field, key resource.PropertyKey, prop resource.PropertyValue,
+	oldInputs, inputs resource.PropertyMap, isCreate bool,
+) resource.PropertyValue {
+	// If the value is already computed or if it is guaranteed to be known, we don't need to do anything
+	if field.known || prop.IsComputed() {
+		return prop
+	}
 
-	// Create mandatory secrets
-	for s := range g.alwaysSecret {
-		s := resource.PropertyKey(s)
-		if m[s].IsComputed() {
-			break
-		}
-		if !m[s].IsSecret() {
-			v := m[s]
-			m[s] = resource.NewSecretProperty(&resource.Secret{Element: v})
+	if input, ok := inputs[key]; ok && !input.IsComputed() && input.DeepEquals(prop) {
+		// prop is an output during a create, but the output mirrors an
+		// input in name and value. We don't make it computed.
+		return prop
+	}
+
+	// If this is during a create and the value is not explicitly marked as known, we mark it computed.
+	if isCreate {
+		return resource.MakeComputed(prop)
+	}
+
+	// If a dependency is computed or has changed, we mark this field as computed.
+	for _, k := range field.deps {
+		k := resource.PropertyKey(k)
+
+		// Not all resources embed their inputs as outputs. When they don't we are
+		// unable to perform old-vs-new diffing here.
+		//
+		// This may lead to applies running on old information during
+		// preview. This is possible anyway, if user's dependencies don't
+		// accurately reflect their logic. This is not a problem for non-preview
+		// updates.
+		//
+		// The solution is to require embedding input structs in output structs
+		// (or do it for the user), ensuring that we have access to information
+		// that changed..
```

Comment:
This PR does not exacerbate this problem. Diffing inputs is currently impossible unless they are embedded, since we don't preserve the old inputs.